### ISSUE TITLE
nng_close deadlock with bus socket

### ISSUE THREAD

**Issue by enricoschroeder** [nanomsg/nng#1247]
This is a follow-up on #1219. 

Even after the recent bugfix commit b7a11a0, nng_close still deadlocks occasionally when using it in conjunction with bus sockets. Below is the simplest way to reproduce the issue, however it occurs under other circumstances as well.

## NNG & Platform details.
nng 1.3.0 or master b7a11a0 (which includes some fixes for nng_close related issues).
Windows (6-core Intel i7-8750H)

## Expected Behavior
nng_close() should close the socket and return.

## Actual Behavior
nng_close() hangs and does not return.

## Steps to Reproduce
Launching two processes that immediately create a raw bus socket each and try to nng_listen() to the same tcp address. One of the nng_listen() calls fails with NNG_EADDRINUSE (this is expected since only one of the nng_listen() calls should succeed). Calling nng_close() on the socket for which the listen call failed.

**Comment by gdamore** [comment]
Seems like a bug.  Will investigate.

**Comment by gdamore** [comment]
Are you using IPC or TCP?

**Comment by gdamore** [comment]
I see you said TCP.

**Comment by gdamore** [comment]
I wrote a test case, but I cannot reproduce the bug.  Here's a sample test program:

```
//
// Copyright 2020 Staysail Systems, Inc. <info@staysail.tech>
//
// This software is supplied under the terms of the MIT License, a
// copy of which should be located in the distribution where this
// file was obtained (LICENSE.txt).  A copy of the license may also be
// found online at https://opensource.org/licenses/MIT.
//

#include <string.h>

#include <nng/nng.h>
#include <nng/protocol/pipeline0/pull.h>
#include <nng/protocol/pipeline0/push.h>
#include <nng/supplemental/util/platform.h>

#include "acutest.h"
#include "testutil.h"

void
test_bug1247(void)
{
	nng_socket bus1, bus2;
	char       addr[64];
	int        rv;

	testutil_scratch_addr("tcp", sizeof(addr), addr);

	TEST_NNG_PASS(nng_bus0_open(&bus1));
	TEST_NNG_PASS(nng_bus0_open(&bus2));

	TEST_NNG_PASS(nng_listen(bus1, addr, NULL, 0));
	rv = nng_listen(bus2, addr, NULL, 0);
	TEST_CHECK(rv != 0);

	TEST_NNG_PASS(nng_close(bus2));
	TEST_NNG_PASS(nng_close(bus1));
}

TEST_LIST = {
	{ "bug1247", test_bug1247 },
	{ NULL, NULL },
};
```

**Comment by gdamore** [comment]
That test code is built using the testutil.c and testutil.h that is part of my standard test suites.  It does not fail.  I'm testing on Windows 10.  I've run this about a dozen times.  I also tried to add a sleep before the closes.  

To diagnose further I need sample test code.

**Comment by gdamore** [comment]
The above test case was merged into the tree proper.  Feel free to play with it to see if you can reproduce.

**Comment by gdamore** [comment]
Its nearly two weeks... I'll probably close this as not reproducible without a response one way or the other soon.

**Comment by enricoschroeder** [comment]
I did not yet have the time to investigate further. It's on my agenda but I don't know when I will get the time.

**Comment by enricoschroeder** [comment]
Ok, I distilled one of the scenarios in our application where the issue occurs into a testcase. 

The offending nng_close call happens inside a pipe notification callback. I now realize that this might be misuse, however in the docs I could not find anything as to what is and what is not allowed inside a pipe notification callback. 

Anyway, here is the test case to reproduce the issue (in case it is indeed correct usage):

```#include <string.h>

#include <nng/nng.h>
#include <nng/protocol/bus0/bus.h>

#include "acutest.h"
#include "testutil.h"

char addr[64];

typedef struct pipe_callback_data {
	nng_socket listener;
	int callback_finished;
} pipe_callback_data_t;

void pipe_callback_fn (nng_pipe pipe, nng_pipe_ev ev, void *userdata)
{
	pipe_callback_data_t *callback_data = (pipe_callback_data_t *) userdata;
	TEST_NNG_PASS(nng_bus0_open_raw(&callback_data->listener));
	if (nng_listen(callback_data->listener, addr, NULL, 0) == NNG_EADDRINUSE) 
	{
		// this close call will hang
		TEST_NNG_PASS(nng_close(callback_data->listener));
	}
	callback_data->callback_finished = 1;
}

void
test_bug1247(void)
{
	nng_socket  listener1; 
	nng_socket  client1, client2;
	nng_dialer  dialer1, dialer2;
	pipe_callback_data_t listener2, listener3;
	memset(&listener2, 0, sizeof(pipe_callback_data_t));
	memset(&listener3, 0, sizeof(pipe_callback_data_t));

	testutil_scratch_addr("tcp", sizeof(addr), addr);

	// init bus listener
	TEST_NNG_PASS(nng_bus0_open_raw(&listener1));
	TEST_NNG_PASS(nng_listen(listener1, addr, NULL, 0));

	// init two bus clients
	TEST_NNG_PASS(nng_bus0_open_raw(&client1));
	TEST_NNG_PASS(nng_pipe_notify(client1, NNG_PIPE_EV_REM_POST, &pipe_callback_fn, &listener2));
	TEST_NNG_PASS(nng_dial(client1, addr, &dialer1, NNG_FLAG_NONBLOCK));
	TEST_NNG_PASS(nng_bus0_open_raw(&client2));
	TEST_NNG_PASS(nng_pipe_notify(client2, NNG_PIPE_EV_REM_POST, &pipe_callback_fn, &listener3));
	TEST_NNG_PASS(nng_dial(client2, addr, &dialer2, NNG_FLAG_NONBLOCK));
	testutil_sleep(100);

	// close the bus listener
	// (this triggers the pipe notification callback where both clients will try to create a new listener) 
	TEST_NNG_PASS(nng_close(listener1));
	testutil_sleep(1000);

	// after some time, both callbacks should have finished executing
	TEST_ASSERT(listener2.callback_finished == 1);
	TEST_ASSERT(listener3.callback_finished == 1);
}

TEST_LIST = {
	{ "bug1247", test_bug1247 },
	{ NULL, NULL },
};
```

**Comment by gdamore** [comment]
Whoa.  Yeah, the only thing you're meant to be able to close at that time is the pipe itself, not the listener or the socket.

There will definitely be a dead lock there.  I'll probably document that.  You're also attempting to open a socket from the pipe event, also a no-no.  Those pipe event callbacks are meant to be able to provide a little bit of connectivity information, not intended to drive orchestration like you're trying to do.

**Comment by enricoschroeder** [comment]
Ok thanks.

We have another similar issue but this might as well be related to the same wrong usage of pipe callbacks on our side. I will refactor it and then report back.

**Comment by enricoschroeder** [comment]
Alright, I refactored our code to not do anything evil from inside the pipe callback and that got rid of the bug as described above.

We had another scenario where nng_close hung, however this was more complicated: It happened under certain circumstances in a Nodejs binding for our software component that uses nng. I suspect it's got something to do with the peculiarities of object lifetime in Nodejs and I got around it by using an updated Nodejs native API that changes this behavior. I could not reproduce the bug outside of Nodejs so I'm assuming for now that it was not a nng bug. For future reference, the location where it hung was here: https://github.com/nanomsg/nng/blob/750662d4aab305d8a3d48bfa6edfc4dac4018881/src/core/socket.c#L754

Anyways, thanks for the help. I'm closing this now.
